### PR TITLE
chore(executor): make allow_missing_host_functions could be configurable

### PR DESCRIPTION
**Pull request by yjhmelody**
We need it, because we wanna set `allow_missing_host_functions`.

### CODE REVIEW

File: client/executor/src/native_executor.rs [hunk 1]
```diff
@@ -170,7 +176,7 @@ where
 	/// runtime is invalidated on any `panic!` to prevent a poisoned state. `ext` is already
 	/// implicitly handled as unwind safe, as we store it in a global variable while executing the
 	/// native runtime.
-	fn with_instance<R, F>(
+	pub fn with_instance<R, F>(
```
Comment (bkchr) [inline client/executor/src/native_executor.rs:179]:
```suggestion
	fn with_instance<R, F>(
```

**Review by bkchr — APPROVED** [review]

File: client/executor/src/native_executor.rs [hunk 2]
```diff
@@ -170,7 +176,7 @@ where
 	/// runtime is invalidated on any `panic!` to prevent a poisoned state. `ext` is already
 	/// implicitly handled as unwind safe, as we store it in a global variable while executing the
 	/// native runtime.
-	fn with_instance<R, F>(
```
Comment (koute) [inline client/executor/src/native_executor.rs:173]: You've added `allow_missing_host_functions` to `self` while this function (and other ones) still take `allow_missing_host_functions` as an argument.

Comment (yjhmelody) [reply client/executor/src/native_executor.rs:173]: Yeah, I fix it.

**Comment by yjhmelody** [comment]
Maybe should I add a method to directly set `allow_missing_host_functions` to be true rather than pass it by `new`?

File: client/executor/src/native_executor.rs (hunk 1 above)
Comment (yjhmelody) [reply client/executor/src/native_executor.rs:179]: To be honest, I think this method should be exported.

**Comment by koute** [comment]
> Maybe should I add a method to directly set `allow_missing_host_functions` to be true rather than pass it by `new`?
> 
> Add extra argument make codebase changed a lot.

Honestly, once you have this many arguments it's probably the best to just make a proper builder instead of jamming everything into `new`. Unless it's actually necessary I don't think adding an extra method to change this argument at runtime is a great idea just for a little convenience's sake.

Anyway, I'm fine with just adding an extra argument to `new`; I'll be refactoring this code anyway in the near future.

File: client/executor/src/native_executor.rs (hunk 1 above)
Comment (bkchr) [reply client/executor/src/native_executor.rs:179]: Why?

**Comment by bkchr** [comment]
> > Maybe should I add a method to directly set `allow_missing_host_functions` to be true rather than pass it by `new`?
> > Add extra argument make codebase changed a lot.
> 
> Honestly, once you have this many arguments it's probably the best to just make a proper builder instead of jamming everything into `new`. Unless it's actually necessary I don't think adding an extra method to change this argument at runtime is a great idea just for a little convenience's sake.
> 
> Anyway, I'm fine with just adding an extra argument to `new`; I'll be refactoring this code anyway in the near future.

Exactly my thinking as well ;)

File: client/executor/src/native_executor.rs (hunk 1 above)
Comment (yjhmelody) [reply client/executor/src/native_executor.rs:179]: Because `CodeExecutor::call` is not generic. I means it use `preregister_builtin_ext`

**Comment by yjhmelody** [comment]
Yeah, I think refactoring is good idea.

**Comment by bkchr** [comment]
@yjhmelody this requires a companion for Polkadot and probably cumulus

**Comment by yjhmelody** [comment]
> this requires a companion for Polkadot and probably cumulus

So if I just add this code:
``` rust
pub fn allow_missing_host_functions(mut self, allow_missing_host_functions: bool) -> Self {
		self.allow_missing_host_functions = allow_missing_host_functions;

		self
	}
```

No need to companion PRs.

And then  maybe @koute could refactor code.

**Comment by bkchr** [comment]
> > this requires a companion for Polkadot and probably cumulus
> 
> So if I just add this code:
> 
> ```rust
> pub fn allow_missing_host_functions(mut self, allow_missing_host_functions: bool) -> Self {
> 		self.allow_missing_host_functions = allow_missing_host_functions;
> 
> 		self
> 	}
> ```
> 
> No need to make companion PRs.
> 
> And then maybe @koute could refactor code.
> 
> So can I change it this way? It does present minimal code changes and is convenient @bkchr .

Sounds good to me.

**Comment by yjhmelody** [comment]
Done!

File: client/executor/src/native_executor.rs [hunk 3]
```diff
@@ -408,13 +423,15 @@ impl<D: NativeExecutionDispatch> NativeElseWasmExecutor<D> {
 	/// `max_runtime_instances` - The number of runtime instances to keep in memory ready for reuse.
 	///
 	/// `runtime_cache_size` - The capacity of runtime cache.
+	///
+	/// `allow_missing_func_imports` - Ignore missing function imports.
```
Comment (koute) [inline client/executor/src/native_executor.rs:427]: You rolled back the parameter but you didn't roll back the comment.

File: client/executor/src/native_executor.rs [hunk 4]
```diff
@@ -137,6 +139,8 @@ where
 	///   compiled execution method is used.
 	///
 	/// `runtime_cache_size` - The capacity of runtime cache.
+	///
+	/// `allow_missing_func_imports` - Ignore missing function imports.
```
Comment (koute) [inline client/executor/src/native_executor.rs:143]: You rolled back the parameter but you didn't roll back the comment.

**Review by koute — APPROVED** [review]
Eh, okay, *fine*. Considering I'll be rewriting all of this code relatively soon I'll let it pass.

File: client/executor/src/native_executor.rs (hunk 3 above)
Comment (yjhmelody) [reply client/executor/src/native_executor.rs]: Ok, fixed.

File: client/executor/src/native_executor.rs (hunk 2 above)
Comment (bkchr) [reply client/executor/src/native_executor.rs:173]: This is still not fixed?

File: client/executor/src/native_executor.rs (hunk 1 above)
Comment (bkchr) [reply client/executor/src/native_executor.rs:181]: Can you please write down exactly on what you want. Why do you need to have this exported. Please write some reasoning. And not just "I means it use preregister_builtin_ext". While this is right, it doesn't explain on what you try to do and fail with the current implementation.

File: client/executor/src/native_executor.rs (hunk 2 above)
Comment (yjhmelody) [reply client/executor/src/native_executor.rs:173]: Sorry, fixed now.

File: client/executor/src/native_executor.rs (hunk 1 above)
Comment (yjhmelody) [reply client/executor/src/native_executor.rs:181]: This `call` in `CodeExecutor` implemented by `WasmExecutor` maybe need to be reimplmented in the future. So we need  to `with_instance` to be `pub`.

Comment (bkchr) [reply client/executor/src/native_executor.rs:181]: > This `call` in `CodeExecutor` implemented by `WasmExecutor` maybe need to be reimplmented in the future.

Why does it needs to be reimplemented?

Comment (yjhmelody) [reply client/executor/src/native_executor.rs:181]: Seems to say it again :) because we want to remove some host functions.

Comment (yjhmelody) [reply client/executor/src/native_executor.rs:181]: Instead, I wonder why these things don't need to be exposed, 
e.g. I think `RuntimeCache` is also very useful.

**Comment by bkchr** [comment]
bot merge

File: client/executor/src/native_executor.rs (hunk 1 above)
Comment (koute) [reply client/executor/src/native_executor.rs:181]: > e.g. I think `RuntimeCache` is also very useful.

FYI, that type will be most likely removed in the refactoring anyway in favor of a more holistic approach.

**Comment by yjhmelody** [comment]
What's wrong with the bot?

**Comment by bkchr** [comment]
bot merge

**Pull request merged**